### PR TITLE
Add positional_encodings_ layer to Dlib

### PR DESCRIPTION
**Pull request by Cydral**
This pull request introduces a new layer, positional_encodings_, to the Dlib library.
The positional_encodings_ layer adds "positional encodings" to the input tensor, which is particularly useful for models processing sequential data, such as transformers. The positional encodings are computed using sine and cosine functions of different frequencies, as described in the paper "Attention is All You Need" by Vaswani et al. This enhancement aims to provide positional information to the model, improving its ability to understand the order and position of elements in a sequence.

The implementation includes methods for setup, forward propagation, and backward propagation, along with serialization support.

### CODE REVIEW

**Comment by pfeatherstone** [comment]
@Cydral This is all great. Really cool. Can I ask though, why do you use dlib instead of Pytorch for neural nets? Slowly adding transformer support to dlib is a lot of work due to the absence of autograd. Its recursive template API makes it very slow to compile. In torch, adding a SOTA layer can be just a few lines of code. For example, sinusoidal positional embeddings is like 15 lines. You can then train in pytorch, then export, or even compile to something you can run in C/C++. So I'm not sure what's the benefit of adding transformer support to dlib when there are arguably better solutions for DNNs in C++. Also, the author has the burden of maintaining this code. I'm not trying to overly criticise but this is requiring a lot of work for potentially little gain. I'm wondering if these additions belong in a "dlib_contrib" repository.

**Comment by Cydral** [comment]
> @Cydral This is all great. Really cool. Can I ask though, why do you use dlib instead of Pytorch for neural nets? Slowly adding transformer support to dlib is a lot of work due to the absence of autograd. Its recursive template API makes it very slow to compile. In torch, adding a SOTA layer can be just a few lines of code. For example, sinusoidal positional embeddings is like 15 lines. You can then train in pytorch, then export, or even compile to something you can run in C/C++. So I'm not sure what's the benefit of adding transformer support to dlib when there are arguably better solutions for DNNs in C++. Also, the author has the burden of maintaining this code. I'm not trying to overly criticise but this is requiring a lot of work for potentially little gain. I'm wondering if these additions belong in a "dlib_contrib" repository.

It's obviously an interesting question and yes, I can confirm that it's also a huge job to add such layers to Dlib, but the idea here isn't necessarily to build a neural network that's exactly the same as what you'll find in other libraries. For example, I'm studying in parallel the impact of convolution layers to replace certain ‘linear’ type layers that you'll actually find in PyTorch.
Of course, the absence of an autograd function also means that we have to design derivatives specifically for the gradient, but it's also a way of getting a better handle on what we're doing. So far I've managed to achieve some interesting results (we'd been trying for a while to introduce a mechanism for predicting the ‘next’ sequence in Dlib using a feed forward principle).
In particular, I've managed to get the network to memorise sequences of text (a mechanism akin to ‘rote’ learning), which I think is already a good start, and even if I decide not to go to the end of this "exercise", the new layers I've added could also be used for specific processing on images too.
In this respect, I think we have a lot of great things to do with Dlib. This isn't the first time I've read about developers moving on to PyTorch, but if nobody tries to develop Dlib any more, then let's shut down development of this library... even though it's still very efficient for many tasks. Adding new functions does not, in principle, oblige the developer to maintain it all. It's also the role of the community around this library to help fix any functional problems it sees (as I sometimes do for certain parts contributed by other contributors).

**Comment by arrufat** [comment]
Yes, I want to encourage you to keep up with this great work! I can't wait to try training transformer-based networks in dlib :)

And of course, I will try my best to help to maintain this stuff :D

**Comment by Cydral** [comment]
I'm continuing and making progress... I've just finished reworking gemm() to take into account the matricial dimension of 4D tensors.

**Comment by davisking** [comment]
Yeah it's all good.  I am a huge fan of pytorch, but there are also things about it I would do differently.  Frankly, if I wasn't extremely busy working at a pre-money startup and trying to ensure we are successful long term (which I feel pretty good about our future), I would be working way more on dlib and making more open source ML stuff in particular.

There is also way more going on in dlib than the deep learning stuff.  The deep learning tooling is honestly the least interesting part of dlib for me.  There are some pretty sweeping changes I would make to the deep learning stuff and will at some point.  Anyway, that's all to say, it doesn't matter how many people use the deep learning parts of dlib.  There are tons of other things in it that lots of people use on a huge number of projects too.

And the dnn tooling is very low dependency and easy to compile, which is really it's selling point.  And people use that still and that's fine.

That's all to say, this PR is cool.  So knock yourself out :)

**Comment by arrufat** [comment]
You should be able to merge the master branch into this one and get the `tril_` layer here too, so the tests pass. Something seems off, where you have the tests of that layer, but not the implementation...

**Comment by Cydral** [comment]
Something went wrong when I retrieved the latest versions after the integration of the tril_ layer... I just deleted the accesses to tril_ in the positional_encodings branch, in the dnn.ccp file. I hope that when the merge is done, nothing will disappear in the master branch.

File: dlib/test/dnn.cpp [hunk 1]
```diff
@@ -4453,47 +4486,6 @@ namespace
...
-            expected_output.host()[tensor_index(expected_output, ii, 0, 0, 1)] = -std::numeric_limits<float>::infinity();
-            expected_output.host()[tensor_index(expected_output, ii, 0, 0, 2)] = -std::numeric_limits<float>::infinity();
-            expected_output.host()[tensor_index(expected_output, ii, 0, 1, 2)] = -std::numeric_limits<float>::infinity();
-        }
-
-        // Compare output tensor with expected output
-        auto& net_output = layer<tag1>(net).get_output();
-        DLIB_TEST(max(abs(mat(net_output) - mat(expected_output))) < 1e-5);
-    }
-
```
Comment (arrufat) [inline dlib/test/dnn.cpp:4496]: This will disappear from master when merging.

Comment (Cydral) [reply dlib/test/dnn.cpp:4496]: Damn, the problem is that I can't manage to merge the master branch in my fork branches, which contains tril_... if only the tril_ tests are missing from dnn.cpp, I can put them back once all the synchronisation has been done.

Comment (Cydral) [reply dlib/test/dnn.cpp]: As I can't get the master version at Dlib level (i.e. with the recently integrated tril_ class), and after resolving the conflicts, I'm now trying to add the definitions for tril_ once again. If the tests pass, we'll have to merge to see if everything works correctly. If it's confirmed, I'll do the same for all the other PRs.

Comment (Cydral) [reply dlib/test/dnn.cpp]: Well, that seems to work. So I'm going to do the same with the other PRs

Comment (davisking) [reply dlib/test/dnn.cpp]: Yeah I merge PRs by squashing the PR into one commit and adding it to master, so master will always be fine regardless.   You can (and probably should here since things have gotten complicated) squash your branch into one commit too.  Having merge commits when you are working with multiple branches is a huge pain, as you are experiencing :D  

There are a bunch of ways to squash a branch into one commit.  I like to use `git reset --soft` to just undo all the commits and then recommit them.  But there is `git merge --squash` and probably several other ways.  Make a backup branch before you do anything though if you aren't real familiar with what I'm talking about though.

Comment (Cydral) [reply dlib/test/dnn.cpp]: OK, but normally you can proceed because I've manually caught up the ‘mixed’ code from the different branches. When everything has been placed under the master branch, I'll remove the branches from my forks to resynchronise the fork and start again from a clean base. All that's left is to rework the last PR for softmaxm, which I'll do today.

Comment (Cydral) [reply dlib/test/dnn.cpp]: @davis, all the "alignments" have been made to enable the merge to go ahead.

**Comment by Cydral** [comment]
@davis, There were indeed new conflicts following the integration of the last class that you dealt with. It looks good now, but could you please go through the integration in chronological order and thus consider frist this new class instead. If it's OK and integrated, I'll make the following changes to avoid going back and forth and I'll keep you informed. Thank you in advance for your help and support.

**Comment by davisking** [comment]
Ah you have a merge error or something.  Check the PR contents, it's missing all your code changes :(

**Comment by Cydral** [comment]
Sorry for all these troubles, I've never had so many problems with merges on GitHub... I've just merged with the main branch and in my session I can see all the classes added recently (transpose, embeddings, tril, ..., and positional_encodings). Could you please have another look from your side? I can't see what code is missing now.

**Comment by davisking** [comment]
> Sorry for all these troubles, I've never had so many problems with merges on GitHub... I've just merged with the main branch and in my session I can see all the classes added recently (transpose, embeddings, tril, ..., and positional_encodings). Could you please have another look from your side? I can't see what code is missing now.

No worries.  Check it again though.  Like look at https://github.com/davisking/dlib/pull/3019/files, it's still missing the changes :shrug:

**Comment by Cydral** [comment]
It's OK now. The implementation in <layers.h> was in fact present but came from a previous merge (which explains why the tests and precompilation worked already)...

File: dlib/dnn/layers_abstract.h [hunk 2]
```diff
@@ -4089,7 +4089,152 @@ namespace dlib
 
 // ----------------------------------------------------------------------------------------
 
-}
+    class positional_encodings_
```
Comment (davisking) [inline dlib/dnn/layers_abstract.h:4092]: Ha yeah wait a minute.  This class is already in this file.  So this duplicates it.  You added it in https://github.com/davisking/dlib/pull/3021.

It looks like all the changes in this PR right now are this extra copy and some formatting and moving stuff around.  I.e. no actual changes.  Maybe we abandon this PR since you already landed what you want in that other PR?   I feel like there was some change in here that was new that you wanted to land in the beginning, but now I forget what it was and it's been lost?
